cellpack/autopack/writers/ImageWriter.py

```diff
@@ -0,0 +1,256 @@
+        )
+        psf = numpy.exp(-(x**2 + y**2 + z**2) / (2 * sigma**2))
+        psf /= psf.sum()
+
+        return psf
+
+    def create_box_psf(self, size=None):
+        """
+        Creates a box psf
+
+        Parameters
+        ----------
+        size: np.ndarray
+            size of the psf
+
+        Returns
+        ----------
+        numpy.ndarray
+            psf
+        """
+        if size is None:
+            size = [3, 3, 3]
+
+        psf = numpy.ones(size)
+        psf /= psf.sum()
+
+        return psf
+
+    def convolve_channel(self, channel, psf):
+        """
+        Convolves a channel with a psf
+
+        Parameters
+        ----------
+        channel: numpy.ndarray
+            channel to be convolved
+        psf: numpy.ndarray
+            psf
+
+        Returns
+        ----------
+        numpy.ndarray
+            convolved channel
+        """
+        scaled_channel = channel.astype(numpy.float32) / 255.0
+        conv_channel = convolve(scaled_channel, psf, mode="constant", cval=0.0)
+        conv_channel = (conv_channel * 255).astype(numpy.uint8)
+        return conv_channel
+
+    def convolve_image(self, image, psf="gaussian", psf_parameters=None):
+        """
+        Convolves the image with a psf
+
+        Parameters
+        ----------
+        image: numpy.ndarray
+            image to be convolved
+        psf: str or numpy.ndarray
+            psf type
+        psf_parameters: dict
+            psf parameters
+
+        Returns
+        ----------
+        numpy.ndarray
+            convolved image
+        """
+        # TODO: add checking for psf_parameters
+        if psf_parameters is None:
+            psf_parameters = {}
+            psf_parameters["sigma"] = 1.5
+            psf_parameters["size"] = [3, 3, 3]
+
+        if isinstance(psf, str):
+            if psf == "gaussian":
+                psf = self.create_gaussian_psf(**psf_parameters)
+            elif psf == "box":
+                psf = self.create_box_psf(psf_parameters.get("size", None))
+            else:
+                raise NotImplementedError(f"PSF type {psf} not implemented")
+
+        conv_img = numpy.zeros(image.shape, dtype=image.dtype)
+        for channel in range(image.shape[0]):
+            conv_img[channel] = self.convolve_channel(image[channel], psf)
+        return conv_img
+
+    @staticmethod
+    def transpose_image_for_projection(image, projection_axis):
+        if projection_axis == "x":
+            image = numpy.transpose(image, axes=(1, 0, 3, 2))
+        elif projection_axis == "y":
+            image = numpy.transpose(image, axes=(2, 0, 3, 1))
+        elif projection_axis == "z":
+            image = numpy.transpose(image, axes=(3, 0, 2, 1))
+        return image
+
+    def create_voxelization(self):
+        """
+        Creates a voxelized representation of the current scene
+        """
+        channel_colors = []
+
+        for pos, rot, ingr, _ in self.env.molecules:
```

Comment:
these two for loops (over molecules and compartments) might be more properly class members of whatever type self.env is.  Then you could call self.env.collect_voxelizations(self.image_data) and the function would fill in the image_data object.